src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -95651,6 +95648,241 @@ class Box<T>
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (12,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(x.Length); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(12, 31),
+                    // (13,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(y.Length); // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(13, 31));
+        }
+
+        [Fact, WorkItem(34942, "https://github.com/dotnet/roslyn/issues/34942")]
+        public void ConditionalAccess_06()
+        {
+            var source =
+@"#nullable enable
+using System;
+
+public class Program
+{
+    static void Main()
+    {
+        string? x = ""42"";
+        string? y = ""43"";
+        if (x?.Length == 2 ? y?.Length == 2 : y?.Length == 3)
+        {
+            Console.WriteLine(x.Length); // 1
+            Console.WriteLine(y.Length);
+        }
+        else
+        {
+            Console.WriteLine(x.Length); // 2
+            Console.WriteLine(y.Length); // 3
+        }
+    }
+}";
+            // The first warning produced for y.Length is unexpected.
+            // https://github.com/dotnet/roslyn/issues/36096
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (12,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(x.Length); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(12, 31),
+                    // (13,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(y.Length);
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(13, 31),
+                    // (17,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(x.Length); // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(17, 31),
+                    // (18,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(y.Length); // 3
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(18, 31)
+);
+        }
+
+        [Fact, WorkItem(34942, "https://github.com/dotnet/roslyn/issues/34942")]
+        public void ConditionalAccess_07()
+        {
+            var source =
+@"#nullable enable
+using System;
+
+public class A
+{
+    public A? P => null;
+
+    public static explicit operator B(A? a) => new B();
+}
+
+public class B { }
+
+
+public class Program
+{
+    static void Main()
+    {
+        A a = new A();
+        B b = new B();
+        if ((B)a.P == b)
```

Comment:
>if ((B)a.P == b) [](start = 8, length = 16)

Consider testing the same with an implicit conversion operator and without the explicit cast.
Also, I'm not sure why you need a property. It would be more direct to test `if (maybeNullA == b) ...` where `maybeNullA` is a parameter of type `A?`.

Reply:
I will simplify to just use a maybe-null parameter.

However I think I can't test aimplicit user-defined conversion the way you describe. Quoting my comment from earlier:

> It looks like a comparison between values of type A and B will not cause a user-defined implicit conversion to occur--you have to cast. Will add as similar of a test to what you've described as possible.

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoUnAZgAJ8qBBKgb1KtZupwIDYq0wAHDGgDGaGFQD2/RBBgSEVAEIAKBhACUVALwA+KgDs4AdyXL1AblIBfUuQ54lTa7ZKUaDgMJOSbdjQAsVACyZt6+vmraBgCuGBiWPuGsilTAUYYmKhYsSbwAZlTKENpaqZqMVDlJNolJVeFoBcrKiurFWqXA5fW+Nb3WQA===